plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -504,221 +532,833 @@ synapticsmst_device_get_flash_checksum (SynapticsMSTDevice *device,
+			return FALSE;
+		}
+	}
+	/* Firmware Checksum */
+	if (chipType == TRUE) {
+		guint32 i =0;
+		for( i = 0; i < code_size + 1; i++ )
+		{
+			checksum += *(payload_data + offset + 0x10 + i);
+		}
+		if ((checksum & 0xFFFF)!=((*(payload_data + 0x408) << 8) | (*(payload_data + 0x409)))) {
+			g_set_error_literal (error,
+						G_IO_ERROR,
+						G_IO_ERROR_INVALID_DATA,
+						"firmware checksum error");
+			return FALSE;
+		}
+		checksum = 0;
+		offset = EEPROM_ESM_OFFSET;
+		for (i = 0; i < (payload_len - EEPROM_ESM_OFFSET); i++)
+		{
+			checksum += *(payload_data + offset + i);
+		}
+		if ((checksum & 0xFFFF)!=((*(payload_data + 0x40A) << 8) | (*(payload_data + 0x40B)))) {
+			g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+							"ESM firmware checksum error");
+			return FALSE;
+		}		
+	}
+	else{	
+		for (guint32 i = 0; i < (code_size + 17); i++)
+			checksum += *(payload_data + offset + i);
 
-	if (checksum & 0xFF) {
+		if (checksum & 0xFF) {
+			g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+						"firmware checksum error");
+			return FALSE;
+		}
+	}
+	return TRUE;
+}	
+
+gboolean
+synapticsmst_device_write_firmware (SynapticsMSTDevice *device,
+					GBytes *fw,
+				    GFileProgressCallback progress_cb,
+				    gpointer progress_data,
+				    GError **error)
+{
+
+	const guint8 *payload_data;
+	guint32 payload_len, payload_len_max;
+	guint32 checksum = 0;
+	guint32 flash_checksum = 0;
+	guint32 offset = 0;
+	guint32 write_loops = 0;
+	guint32 data_to_write = 0;
+	guint32 dwData[16];
+	guint16 tmp;
+	guint16 erase_code = 0xFFFF;	
+	guint8 percentage = 0;
+	guint8 rc = 0;
+	gboolean isChipPanamera = FALSE;
+	SynapticsMSTDevicePrivate *priv = GET_PRIVATE (device);
+	g_autoptr(SynapticsMSTConnection) connection = NULL;
+
+	payload_data = g_bytes_get_data (fw, NULL);
+	payload_len = g_bytes_get_size (fw);
+
+	if (synapticsmst_device_get_board_id (device)> 0x5000){
+		isChipPanamera = TRUE;
+		payload_len_max = PAYLOAD_SIZE_512K;
+							g_debug("pana");
+
+	}
+	else {
+		isChipPanamera = FALSE;
+		payload_len_max = PAYLOAD_SIZE_64K;		
+	}
+
+	if (!synapticsmst_device_check_firmware_content (device, fw, isChipPanamera, error)){
 		g_set_error_literal (error,
 				     G_IO_ERROR,
 				     G_IO_ERROR_INVALID_DATA,
-				     "firmware checksum error");
+				     "invalid file content");
 		return FALSE;
 	}
+							g_debug("check id");
 
 	/* TODO: May need a way to override this to cover field
 	 * issues of invalid firmware flashed*/
 	/* check firmware and board ID again */
 	tmp = (*(payload_data + ADDR_CUSTOMER_ID) << 8) + *(payload_data + ADDR_BOARD_ID);
-	if (tmp != synapticsmst_device_get_board_id (device)) {
-		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "board ID mismatch");
-		return FALSE;
+	if (isChipPanamera == TRUE){
+		
+		g_debug("bypass check CusID %x %x",tmp, synapticsmst_device_get_board_id (device));	
```

Comment:
This should actually be a check for EVB board not for bypassing customer ID not ignoring customer ID on all panamera boards I'd think.